### DISCUSSION TITLE
RLC series vs parallel + inductor support?

### DISCUSSION THREAD

**Discussion by 0xCoto** [thliebig/openEMS-Project#65] [Q&A Usage]
I'm looking through the docs: https://docs.openems.de/python/CSXCAD/CSProperties/CSPropLumpedElement.html#CSXCAD.CSProperties.CSPropLumpedElement

I think I remember having read somewhere in the old forum a while ago that inductors were not supported in lumped elements yet. Is that still the case, or are RLC lumped elements fully functional now for both resistors, inductors and capacitors?

Also, is it possible to specify RLC series vs RLC parallel?

**Comment by thliebig** [comment]
I think not much has changed on that front. So I think only RC is still the only option...

**Reply by 0xCoto** [reply]
Any idea how much effort it would require to implement inductors too, and whether being able to set RLC parallel instead of the default series could be an option?

**Reply by Sarajevo67** [reply]
**instead of the default series**

I thought that parallel is default (and currently only) RC configuration?

**Reply by 0xCoto** [reply]
That sounds right, my bad. In any case, with the latest PR, both parallel and serial are supported, including inductors - so full RLC rather than just RC.

**Reply by Sarajevo67** [reply]
Really good news, JIT!!! 
I just realized that lumped element RLC-staged impedances, might be the only way to simulate impedance transformer.

**Comment by 0xCoto** [comment]
@thliebig - we're trying to implement inductor support by modifying [`operator.cpp`](https://github.com/thliebig/openEMS/blob/master/FDTD/operator.cpp) like so:

```cpp
bool Operator::Calc_LumpedElements()
{
	vector<CSProperties*> props = CSX->GetPropertyByType(CSProperties::LUMPED_ELEMENT);
	for (size_t i=0;i<props.size();++i)
	{
		CSPropLumpedElement* PLE = dynamic_cast<CSPropLumpedElement*>(props.at(i));
		if (PLE==NULL)
			return false; //sanity check: this should never happen!
		vector<CSPrimitives*> prims = PLE->GetAllPrimitives();
		for (size_t bn=0;bn<prims.size();++bn)
		{
			CSPrimBox* box = dynamic_cast<CSPrimBox*>(prims.at(bn));
			if (box)
			{	//calculate lumped element parameter

				double C = PLE->GetCapacity();
				if (C<=0)
					C = NAN;
				double R = PLE->GetResistance();
				if (R<0)
					R = NAN;
				
				// Add lumped inductor
				double L = PLE->GetInductance();
				if (L<=0)
					L = NAN;
				// End addition
				
				if ((std::isnan(R)) && (std::isnan(C)))
				{
					cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element R or C not specified! skipping. "
							<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
					continue;
				}

				int ny = PLE->GetDirection();
				if ((ny<0) || (ny>2))
				{
					cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element direction is invalid! skipping. "
							<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
					continue;
				}
				int nyP = (ny+1)%3;
				int nyPP = (ny+2)%3;

				unsigned int uiStart[3];
				unsigned int uiStop[3];
				// snap to the native coordinate system
				int Snap_Dimension = Operator::SnapBox2Mesh(box->GetStartCoord()->GetCoords(m_MeshType), box->GetStopCoord()->GetCoords(m_MeshType), uiStart, uiStop, false, true);
				if (Snap_Dimension<=0)
				{
					if (Snap_Dimension>=-1)
						cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element snapping failed! Dimension is: " << Snap_Dimension << " skipping. "
								<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
					// Snap_Dimension == -2 means outside the simulation domain --> no special warning, but box probably marked as unused!
					continue;
				}

				if (uiStart[ny]==uiStop[ny])
				{
					cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element with zero (snapped) length is invalid! skipping. "
							<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
					continue;
				}

				//calculate geometric property for this lumped element
				unsigned int pos[3];
				double unitGC=0;
				// Add container for inductor "area"
				double unitL = 0;
				// End Add
				int ipos=0;
				for (pos[ny]=uiStart[ny];pos[ny]<uiStop[ny];++pos[ny])
				{
					double unitGC_Plane=0;
					// Add container for inductor "area"
					double unitL_plane = 0;
					// End Add
					for (pos[nyP]=uiStart[nyP];pos[nyP]<=uiStop[nyP];++pos[nyP])
					{
						for (pos[nyPP]=uiStart[nyPP];pos[nyPP]<=uiStop[nyPP];++pos[nyPP])
						{
							// capacity/conductivity in parallel: add values
							unitGC_Plane += GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos);
							// Add container for inductor "area"
							if (unitL_plane == 0)
								unitL_plane = 1/(GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos));
							else
								unitL_plane += 1/(GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos));
							// End Add
						}
					}

					//capacity/conductivity in series: add reciprocal values
					unitGC += 1/unitGC_Plane;
					// Calculate addition in series
					unitL += 1/unitL_plane;
					// End Add
				}
				unitGC = 1/unitGC;
				
				bool caps = PLE->GetCaps();
				double kappa = 0;
				double epsilon = 0;
				
				if (R>0)
					kappa = 1 / R / unitGC;
				if (C>0)
				{
					epsilon =  C / unitGC;

					if (epsilon< __EPS0__)
					{
						cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element capacity is too small for its size! skipping. "
								<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
						C = 0;
					}
				}
				
				// Attempt to add mu modification 
				double mu = 0;
				if (L>0)
				{
					mu = L / unitL;
					
					if (mu < __MUE0__)
					{
						cerr << "Operator::Calc_LumpedElements(): Warning: Lumped Element inductance is too small for its size! skipping. "
								<< " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
						L = 0;
					}
				}
				// End Addition
				
				for (pos[ny]=uiStart[ny];pos[ny]<uiStop[ny];++pos[ny])
				{
					for (pos[nyP]=uiStart[nyP];pos[nyP]<=uiStop[nyP];++pos[nyP])
					{
						for (pos[nyPP]=uiStart[nyPP];pos[nyPP]<=uiStop[nyPP];++pos[nyPP])
						{
							ipos = MainOp->SetPos(pos[0],pos[1],pos[2]);
							if (C>0)
								EC_C[ny][ipos] = epsilon * GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos);
							if (R>0)
								EC_G[ny][ipos] = kappa * GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos);
							
							// Attempt to add inductance
							if (L>0)
								EC_L[ny][ipos] = mu * GetEdgeArea(ny,pos)/GetEdgeLength(ny,pos);
							// End addition
							
							if (R==0) //make lumped element a PEC if resistance is zero
							{
								SetVV(ny,pos[0],pos[1],pos[2], 0 );
								SetVI(ny,pos[0],pos[1],pos[2], 0 );
							}
							else //recalculate operator inside the lumped element
								Calc_ECOperatorPos(ny,pos);
						}
					}
				}

				// setup metal caps
				if (caps)
				{
					for (pos[nyP]=uiStart[nyP];pos[nyP]<=uiStop[nyP];++pos[nyP])
					{
						for (pos[nyPP]=uiStart[nyPP];pos[nyPP]<=uiStop[nyPP];++pos[nyPP])
						{
							pos[ny]=uiStart[ny];
							if (pos[nyP]<uiStop[nyP])
							{
								SetVV(nyP,pos[0],pos[1],pos[2], 0 );
								SetVI(nyP,pos[0],pos[1],pos[2], 0 );
								++m_Nr_PEC[nyP];
							}

							if (pos[nyPP]<uiStop[nyPP])
							{
								SetVV(nyPP,pos[0],pos[1],pos[2], 0 );
								SetVI(nyPP,pos[0],pos[1],pos[2], 0 );
								++m_Nr_PEC[nyPP];
							}

							pos[ny]=uiStop[ny];
							if (pos[nyP]<uiStop[nyP])
							{
								SetVV(nyP,pos[0],pos[1],pos[2], 0 );
								SetVI(nyP,pos[0],pos[1],pos[2], 0 );
								++m_Nr_PEC[nyP];
							}

							if (pos[nyPP]<uiStop[nyPP])
							{
								SetVV(nyPP,pos[0],pos[1],pos[2], 0 );
								SetVI(nyPP,pos[0],pos[1],pos[2], 0 );
								++m_Nr_PEC[nyPP];
							}
						}
					}
				}
				box->SetPrimitiveUsed(true);

			}
			else
				cerr << "Operator::Calc_LumpedElements(): Warning: Primitives other than boxes are not supported for lumped elements! skipping "
						<< prims.at(bn)->GetTypeName() << " ID: " << prims.at(bn)->GetID() << " @ Property: " << PLE->GetName() << endl;
		}
	}
	return true;
}
```

However, it seems that adjusting the value of L does not appear to make any influence at all on the results (testing with Python):

```python
lumped_1 = CSX.AddLumpedElement("lumped_1", ny="z", caps=True, R=40, L=1e-5, C=1e-12)
lumped_1.AddBox([0,1.035,-1], [0,1.035,1])
```

I'm basically trying to place the RLC lumped element between the two ends of a microstrip line gap:
<img width="1146" alt="Screenshot 2023-06-02 at 03 36 09" src="https://github.com/thliebig/openEMS-Project/assets/25392776/8a3a884d-5f6b-47a1-9ee9-911efe84690c">

Any ideas why that might be? Would we need to look at any other part of the codebase for `L` to be picked up?

---

P.S.: The good news is that in terms of accuracy, the `R=40, L=1e-5, C=1e-12` configuration (though `L` doesn't seem to make any influence for some reason), appears to present great agreement with CST, so something is definitely working:

<img width="1393" alt="Screenshot 2023-06-02 at 02 01 19" src="https://github.com/thliebig/openEMS-Project/assets/25392776/a999e266-51fb-42e8-9f7f-f8d9ac3a5a10">
<img width="425" alt="Screenshot 2023-06-02 at 02 10 26" src="https://github.com/thliebig/openEMS-Project/assets/25392776/4f64e929-8844-4070-8a35-745bd0e754f1">

**Reply by thliebig** [reply]
Hi, if that all would be as easy as you have attempted it above, I would have done it back in the days directly ;)

The problem is that C and G change/influence the electric fields (voltages in EC-FDTD in openEMS). But the inductance has to work with the magnetic fields (currents in EC-FDTD). 
But the magnetic fields do not reside on the primary mesh, but on the secondary. Additionally you must not change the normal components (in the direction of the "real" currents aka, kappa*E) but the tangential directions where the magnetic fields curling aroung the "conductor". 
All that said it is clear that an inductor has to be handled very differently. I really would have to dive much deeper into the FDTD algorithm again to figure it out. It should not be too hard, but it is certainly not as trivial as you tried above.

But it makes perfectly sense that you did not see any affect as you changed the mue_r in the direction of the electric fields and did not at all change the ones that act on the surrounding magnetic fields...

I hope that all makes sense?

Additionally I really think you should create a better test structure.
I would recommend something like a wire loop with only a single lumped element. Then in a reference simulation you calculate the lumped element  with R=0 (short circuit) to be able to extract the influence of the feeding wire and feeding port.
Then when you simulate the lumped port with finite values you can extract **only** the lumped element values as e.g. Z over f...

best regards
Thorsten

**Comment by biergaizi** [comment]
In my opinion the lack of support for lumped inductance and lumped S-parameters is so far the most significant barrier for electronics design simulations, this must be addressed in the near future before openEMS becomes fully usable for circuit board design. To accomplish the goal, the use of EC-FDTD is a source of difficulty, it's equivalent to FDTD but a different formalism that expresses E and H fields as voltages and currents, so one can't just copy the result from a standard FDTD textbook but it must be rederived. 

I'm a programmer that just happens to tinker with 1 GHz PCBs for fun, I'm not a physicist, without a PhD in physics I'm not qualified to solve this problem. But let's hope for the future...

**Reply by thliebig** [reply]
Well the EC-FDTD approach makes many things much easier. Implementing RLC should be one of them.
Additionally any textbook FDTD should be easily adaptable to EC-FDTD and if that already seems to complicated then it would be a good idea not to even attempt to implement it ;)

**Comment by 0xCoto** [comment]
(We're looking into this - hoping to have a PR of the implementation ready in the upcoming weeks)

**Comment by 0xCoto** [comment]
It seems that our implementation is complete. RLC circuits (parallel as well as serial) are now supported, and we get accurate results. Tomorrow, I'll share the mathematical formulations we went through, as well as some comparisons with Qucs. We'll open a PR in CSXCAD and one in the openEMS repository.

**Comment by 0xCoto** [comment]
@thliebig - Ready! Feel free to review and merge these two PRs whenever you find the time:
- https://github.com/thliebig/CSXCAD/pull/34
- https://github.com/thliebig/openEMS/pull/121

Any questions or comments welcome.
